File: src/vehicle/part/link/SimpleLink.cpp
#include "SimpleLink.h"

void SimpleLink::activate(btRigidBody* from, btTransform from_frame, btRigidBody* to, btTransform to_frame)
{
	if (constraint == nullptr)
	{

		constraint = new btGeneric6DofSpringConstraint(
			*from, *to, from_frame, to_frame, true);


		constraint->enableSpring(2, true);
		constraint->setStiffness(2, 10000.0);
		constraint->setDamping(2, 0.000005);
		constraint->setBreakingImpulseThreshold(30.0);
		constraint->setLinearUpperLimit(btVector3(0., 0., 10.5));
		constraint->setLinearLowerLimit(btVector3(0., 0., -10.5));

		constraint->setAngularLowerLimit(btVector3(0.f, 0.f, -1.5f));
		constraint->setAngularUpperLimit(btVector3(0.f, 0.f, 1.5f));

		constraint->setDbgDrawSize(btScalar(5.f));


		world->addConstraint(constraint, true);


	}
}

void SimpleLink::deactivate()
{
	if (constraint != nullptr)
	{
		world->removeConstraint(constraint);
		delete constraint;

		constraint = nullptr;
	}
}

bool SimpleLink::is_broken()
{
	if (constraint == nullptr)
	{
		return false;
	}

	return !constraint->isEnabled();
}

void SimpleLink::set_breaking_enabled(bool value)
{
	if (constraint)
	{
		constraint->setBreakingImpulseThreshold(value ? 30.0 : 100000000.0);
	}
}

SimpleLink::SimpleLink(btDynamicsWorld* world)
{
	this->world = world;
	constraint = nullptr;
}

SimpleLink::~SimpleLink()
{
	deactivate();
}

File: res/core/links/simple_link.lua
local bullet = require("bullet");
local glm = require("glm");

local constraint = nil;
local bt_world = nil;

function load_toml(toml_data)

end

function activate(from_rigid, from_frame, to_rigid, to_frame, world)
	if constraint == nil then 
		bt_world = world

		constraint = bullet.generic_6dof_spring_constraint.new(from_rigid, to_rigid, from_frame, to_frame, true)
		constraint:enable_spring(2, true)
		constraint:set_stiffness(2, 1000000.0)
		constraint:set_damping(2, 0.00001)
		constraint:set_linear_upper_limit(glm.vec3.new(0.04, 0.04, 0.5))
		constraint:set_linear_lower_limit(glm.vec3.new(-0.04, -0.04, -0.5))
		constraint:set_angular_upper_limit(glm.vec3.new(0, 0, 1.5))
		constraint:set_angular_lower_limit(glm.vec3.new(0, 0, -1.5))
		constraint:set_debug_draw_size(5.0)
		constraint:set_equilibrium_point()

		constraint:enable_spring(0, true)
		constraint:set_stiffness(0, 100000.0)
		constraint:set_damping(0, 0.00001)
		
		constraint:enable_spring(1, true)
		constraint:set_stiffness(1, 100000.0)
		constraint:set_damping(1, 0.00001)


		constraint:add_to_world(world, true)
	end
end

function deactivate()	
	if constraint ~= nil then 
		constraint:remove_from_world(bt_world)
		bt_world = nil
		constraint = nil
	end

end 

function is_broken() 

	if constraint == nil then
		return true 
	else
		return not constraint:is_enabled()
	end

end 

function set_breaking_enabled(value)


end
